juju/juju
# refactor: call ModelConfigAPI instead of ControllerAPI.ModelConfig

This PR is prerequisite work to a change which will come in 4.0. While attempting to move the `Controller` facade over to model services, we realised that the controller facade *should* be a multi-model facade. Currently it is a model-scoped facade. However, we can't change it over right now due to the `ModelConfig` facade method, which requires the facade to be scoped to a single model.

So in this PR, we replace all usages of `Controller.ModelConfig` with the `ModelConfig` facade. This ensures that 3.6 clients will continue to be compatible with 4.0 controllers when we change the `Controller` facade in 4.0.

This PR does not include any facade changes, but we have removed the `ModelConfig` method from the controller API client to discourage people from using that facade method. Furthermore, we have added a deprecation warning to the facade method. In 4.0, we will remove this facade method and bump the facade version.

## Checklist




## QA steps

Affected commands are `upgrade-model`, `destroy-controller`, `kill-controller`, `show-controller`. The following should test all of the code paths that have been modified.

Bootstrap Juju, run `show-controller`, and check that the controller model version displays correctly:
```
$ juju show-controller | grep controller-model-version
    controller-model-version: 3.6-beta2.1
```

Run `upgrade-model` on the controller model, and check that it raises an error:
```
$ juju switch controller
$ juju upgrade-model
ERROR use upgrade-controller to upgrade the controller model
```

Run `destroy-controller` and check that the controller is successfully destroyed.

Bootstrap Juju again, run `kill-controller` and check that the controller is successfully destroyed.

## Links

**Jira card:** JUJU-6350
